File: Entidades/Stock.vb
Public Class Stock

    Private id_Stock_, id_Usuario_, id_Sucursal_, id_Producto_, Stock_Minimo_, Stock_Optimo_, Stock_Actual_, Habilitado_, Modificado_ As Integer
    Private Motivo_, Usuario_ As String
    Private Fecha_ As DateTime
    Private Fecha_Mod_ As Date

    Public Property id_Stock() As Integer
        Get
            Return id_Stock_
        End Get
        Set(ByVal value As Integer)
            id_Stock_ = value
        End Set
    End Property

    Public Property id_Producto() As Integer
        Get
            Return id_Producto_
        End Get
        Set(ByVal value As Integer)
            id_Producto_ = value
        End Set
    End Property

    Public Property id_Sucursal() As Integer
        Get
            Return id_Sucursal_
        End Get
        Set(ByVal value As Integer)
            id_Sucursal_ = value
        End Set
    End Property

    Public Property Stock_Minimo() As Integer
        Get
            Return Stock_Minimo_
        End Get
        Set(ByVal value As Integer)
            Stock_Minimo_ = value
        End Set
    End Property

    Public Property Stock_Actual() As Integer
        Get
            Return Stock_Actual_
        End Get
        Set(ByVal value As Integer)
            Stock_Actual_ = value
        End Set
    End Property

    Public Property Stock_Optimo() As Integer
        Get
            Return Stock_Optimo_
        End Get
        Set(ByVal value As Integer)
            Stock_Optimo_ = value
        End Set
    End Property
    Public Property Fecha() As DateTime
        Get
            Return Fecha_
        End Get
        Set(ByVal value As DateTime)
            Fecha_ = value
        End Set
    End Property

    Public Property Fecha_Mod() As Date
        Get
            Return Fecha_Mod_
        End Get
        Set(ByVal value As Date)
            Fecha_Mod_ = value
        End Set
    End Property

    Public Property Motivo() As String
        Get
            Return Motivo_
        End Get
        Set(ByVal value As String)
            Motivo_ = value
        End Set
    End Property

    Public Property id_Usuario() As Integer
        Get
            Return id_Usuario_
        End Get
        Set(ByVal value As Integer)
            id_Usuario_ = value
        End Set
    End Property

    Public Property Usuario() As String
        Get
            Return Usuario_
        End Get
        Set(ByVal value As String)
            Usuario_ = value
        End Set
    End Property

    Public Property Modificado() As Integer
        Get
            Return Modificado_
        End Get
        Set(ByVal value As Integer)
            Modificado_ = value
        End Set
    End Property

    Public Property Habilitado() As Integer
        Get
            Return Habilitado_
        End Get
        Set(ByVal value As Integer)
            Habilitado_ = value
        End Set
    End Property
End Class

File: Stock/Stock.Core/Model/BaseAgreggate/Stock.cs
using Common.Core.Enum;
using Common.Core.Exceptions;
using Common.Core.Model;
using System;
using System.Linq;

namespace Stock.Core.Model.BaseAgreggate
{
    public class Stock : Entity<long>
    {
        public int IdSucursal { get; private set; }
        public int IdProducto { get; private set; }
        public Producto Producto { get; private set; }
        public int Cantidad { get; private set; }
        public int Minimo { get; private set; }
        public int Optimo { get; private set; }
        public int VentaMensual { get; private set; }
        public int Reservado { get; private set; }
        public int Restante => Cantidad - Reservado;
        public bool Habilitado { get; private set; }
        public string MotivoModificacion { get; private set; }
        public int? IdUsuario { get; private set; }
        public Usuario Usuario { get; private set; }
        public bool Modificado { get; private set; }
        public bool Borrado { get; private set; }
        public DateTime Fecha { get; private set; }
        public DateTime FechaModificacion { get; private set; }
        public DateTime FechaEdicion { get; private set; }

        public Stock()
        {
        }

        public Stock(int idSucursal, Producto producto, int cantidad, int minimo, int optimo, bool habilitado, Usuario usuario) : base(true)
        {
            IdSucursal = idSucursal;
            IdProducto = producto.Id;
            Producto = producto;
            Cantidad = cantidad;
            Minimo = minimo;
            Optimo = optimo;
            IdUsuario = usuario.Id;
            Usuario = usuario;
            Habilitado = habilitado;
            Modificado = false;
            Borrado = false;
            Fecha = DateTime.Now;
            FechaModificacion = DateTime.Now;
            FechaEdicion = DateTime.Now;
            VentaMensual = 0;
        }

        public void Actualiza(int? cantidad, int? minimo, int? optimo, int? ventaMensual, bool? habilitado, string motivo, bool? borrado, Usuario usuario)
        {
            if (minimo > optimo)
                throw new NegocioException("Error al actualizar el stock. El stock minimo no puede ser mayor al optimo.");

            if (string.IsNullOrWhiteSpace(motivo))
                if(!borrado.HasValue)
                    throw new NegocioException("Error al actualizar el stock. Debe ingresar un motivo.");
                else
                    throw new NegocioException("Error al borrar el stock. Debe ingresar un motivo.");

            if(cantidad.HasValue)
                Cantidad = cantidad.Value;

            if (minimo.HasValue)
                Minimo = minimo.Value;

            if (optimo.HasValue)
                Optimo = optimo.Value;

            if (ventaMensual.HasValue)
                VentaMensual = ventaMensual.Value;

            if (habilitado.HasValue)
                Habilitado = habilitado.Value;

            if (borrado.HasValue)
                Borrado = borrado.Value;

            MotivoModificacion = motivo;
            IdUsuario = usuario.Id;
            Usuario = usuario;
            Modificado = true;
            FechaModificacion = DateTime.Now;
            FechaEdicion = DateTime.Now;
        }

        public void CargarReserva(int reserva)
        {
            Reservado = reserva;
        }
    }
}
